core/lib/parameter_hunter.js

```diff
@@ -13,110 +13,227 @@
+   *   * Save the value to the values array.
+   *   * The do-while loop truncates the paramString value while parsing. Its
+   *     condition for completion is when the paramString is whittled down to an
+   *     empty string.
+   *   * After the keys and values arrays are built, a for loop iterates through
+   *     them to build the final paramStringWellFormed string.
+   *   * No quote substitution had been done prior to this loop. In this loop,
+   *     all keys are ensured to be wrapped in double-quotes. String values are
+   *     also ensured to be wrapped in double-quotes.
+   *   * Unescape escaped unicodes except for double-quotes. Everything beside
+   *     double-quotes will be wrapped in double-quotes without need for escape.
+   *   * Return paramStringWellFormed.
+   *
+   * @param {string} pString
+   * @returns {string} paramStringWellFormed
+   */
   function paramToJson(pString) {
-    var paramStringWellFormed = '';
-    var paramStringTmp;
-    var colonPos;
-    var delimitPos;
-    var quotePos;
-    var paramString = pString;
+    var colonPos = -1;
+    var keys = [];
+    var paramString = pString; // to not reassign param
+    var paramStringWellFormed;
+    var quotePos = -1;
+    var regex;
+    var values = [];
+    var wrapper;
 
-    do {
+    //replace all escaped double-quotes with escaped unicode
+    paramString = paramString.replace(/\\"/g, '\\u0022');
 
-      //if param key is wrapped in single quotes, replace with double quotes.
-      paramString = paramString.replace(/(^\s*[\{|\,]\s*)'([^']+)'(\s*\:)/, '$1"$2"$3');
+    //replace all escaped single-quotes with escaped unicode
+    paramString = paramString.replace(/\\'/g, '\\u0027');
 
-      //if params key is not wrapped in any quotes, wrap in double quotes.
-      paramString = paramString.replace(/(^\s*[\{|\,]\s*)([^\s"'\:]+)(\s*\:)/, '$1"$2"$3');
+    //replace all escaped commas with escaped unicode
+    paramString = paramString.replace(/\\,/g, '\\u0044');
 
-      //move param key to paramStringWellFormed var.
-      colonPos = paramString.indexOf(':');
+    //replace all escaped colons with escaped unicode
+    paramString = paramString.replace(/\\:/g, '\\u0058');
 
-      //except to prevent infinite loops.
-      if (colonPos === -1) {
-        colonPos = paramString.length - 1;
-      }
-      else {
-        colonPos += 1;
-      }
-      paramStringWellFormed += paramString.substring(0, colonPos);
-      paramString = paramString.substring(colonPos, paramString.length).trim();
+    //with escaped chars out of the way, crawl through paramString looking for
+    //keys and values
+    do {
 
-      //if param value is wrapped in single quotes, replace with double quotes.
-      if (paramString[0] === '\'') {
-        quotePos = paramString.search(/[^\\]'/);
+      //check if searching for a key
+      if (paramString[0] === '{' || paramString[0] === ',') {
+        paramString = paramString.substring([1], paramString.length).trim();
```

Comment:
The first `substring()` arg should be an integer, not an array (probably just a typo). It happens to work since JS coerces `[1]` to `1` in this case, but still worth tidying up to avoid confusion/bugs in future.